### PR TITLE
Add a cleanup go routine that runs every hour to check to see if ther…

### PR DESCRIPTION
**Pull request by mikeshng**
…e are any orphan subscriptionstatus, if there are delete the resources listed inside of them

Signed-off-by: Mike Ng <ming@redhat.com>

### CODE REVIEW

**Comment by xiangjingli** [comment]
@mikeshng  Thanks Mike! It looks good.  We may let customer know that the clean up happens every hour. It won't happen right away.

@philipwu08  Could you help double review the fix? The fix is per an issue reported in the slack channel. Here is the use case in the hub restore scenario
- customer shut down the managed cluster
- customer switch hub cluster to a new one  
- customer delete the hub appsub or change the placementrule to switch the hub appsub to other managed clusters
- customer bring back the managed cluster

In this case, when the application manager is running again, the appsub deployed on the managed cluster has gone. So there is no chance for application manager to clean up the resources deployed by the deleted appsub.

**Comment by mikeshng** [comment]
@xiangjingli I tested it and the cleanup happened right away on startup of the app mgr addon pod so I think we should be good.

File: pkg/synchronizer/kubernetes/sync_server.go [hunk 1]
```diff
@@ -236,3 +243,56 @@ func (sync *KubeSynchronizer) GetRemoteClient() client.Client {
...
+					if errors.IsNotFound(err) {
+						klog.Infof("cannot find Subscription namespace: %s , name: %s , deleting resources in SubscriptionStatus",
+							nsn.Namespace, nsn.Name)
+
+						if len(appsubStatus.Statuses.SubscriptionStatus) > 0 {
+							foundErr := false
+
+							for _, unitStatus := range appsubStatus.Statuses.SubscriptionStatus {
+								if err = synchronizer.DeleteSingleSubscribedResource(nsn, unitStatus); err != nil {
+									klog.Error(err, "failed to delete resource")
```
Comment (xiangjingli) [inline pkg/synchronizer/kubernetes/sync_server.go:281]: should we assign foudErr = true here?

Comment (mikeshng) [reply pkg/synchronizer/kubernetes/sync_server.go:281]: > should we assign foudErr = true here?

Thanks for the catch! Fixed.

**Comment by mikeshng** [comment]
@philipwu08 hope you feel better. Could you please take a look at this change? Thanks.

File: pkg/synchronizer/kubernetes/sync_server.go [hunk 2]
```diff
@@ -236,3 +243,57 @@ func (sync *KubeSynchronizer) GetRemoteClient() client.Client {
 func (sync *KubeSynchronizer) GetRemoteNonCachedClient() client.Client {
 	return sync.RemoteNonCachedClient
 }
+
+// startCleanup looks up all the subscriptionstatuses
+// For each subscriptionstatus that doesn't have a subscription
+// Delete the resources listed inside the subscriptionstatus
+// If all the resources are deleted successfully then delete the subscriptionstatus
+func startCleanup(synchronizer *KubeSynchronizer) {
```
Comment (BenamarMk) [inline pkg/synchronizer/kubernetes/sync_server.go:251]: This solution works, but I would like to propose an additional improvement. Currently, we have several bugs (BZs) that indicate that the workload is not being deleted until after an hour from the node restart. I have added a comment here to explain this issue [here](https://bugzilla.redhat.com/show_bug.cgi?id=2155472#c10)

So my suggestion is to use an exponential delay that increases up to a maximum of one hour.

Comment (mikeshng) [reply pkg/synchronizer/kubernetes/sync_server.go:251]: @xiangjingli what do you think about @BenamarMk 's suggestion?

Comment (xiangjingli) [reply pkg/synchronizer/kubernetes/sync_server.go:251]: Yes, it is good suggestion. Mike will implement the exponential backoff wait mechanism

Comment (mikeshng) [reply pkg/synchronizer/kubernetes/sync_server.go:251]: @BenamarMk done as suggested in https://github.com/open-cluster-management-io/multicloud-operators-subscription/pull/308

Comment (BenamarMk) [reply pkg/synchronizer/kubernetes/sync_server.go:251]: Thank you @mikeshng

**Pull request merged**